hypothesis-python/src/hypothesis/strategies/_internal/datetime.py

```diff
@@ -62,30 +88,50 @@ def draw_capped_multipart(data, min_value, max_value):
         result[name] = val
         cap_low = cap_low and val == low
         cap_high = cap_high and val == high
+    if hasattr(min_value, "fold"):
+        # The `fold` attribute is ignored in comparison of naive datetimes.
+        # In tz-aware datetimes it would require *very* invasive changes to
+        # the logic above, and be very sensitive to the specific timezone
+        # (at the cost of efficient shrinking and mutation), so at least for
+        # now we stick with the status quo and generate it independently.
+        result["fold"] = utils.integer_range(data, 0, 1)
     return result
 
 
 class DatetimeStrategy(SearchStrategy):
-    def __init__(self, min_value, max_value, timezones_strat):
+    def __init__(self, min_value, max_value, timezones_strat, allow_imaginary):
         assert isinstance(min_value, dt.datetime)
         assert isinstance(max_value, dt.datetime)
         assert min_value.tzinfo is None
         assert max_value.tzinfo is None
         assert min_value <= max_value
         assert isinstance(timezones_strat, SearchStrategy)
+        assert isinstance(allow_imaginary, bool)
         self.min_value = min_value
         self.max_value = max_value
         self.tz_strat = timezones_strat
+        self.allow_imaginary = allow_imaginary
 
     def do_draw(self, data):
-        result = draw_capped_multipart(data, self.min_value, self.max_value)
-        result = dt.datetime(**result)
+        # We start by drawing a timezone, and an initial datetime.
         tz = data.draw(self.tz_strat)
+        result = self.draw_naive_datetime_and_combine(data, tz)
+
+        # TODO: with some probability, systematically search for one of
+        #   - an imaginary time (if allowed),
+        #   - a time within 24hrs of a leap second (if there any are within bounds),
```

Comment:
Obviously this is a placeholder for #69, since all the recent datetimes work has been leading up to that.  In brief,

- leap second issues can appear when comparing durations across a leap second calculated between (e.g. unix) timestamps to those calculated between dates
- integrating a bias towards leap-second smear periods into the primary `datetimes()` strategy can support substantially better mutation and shrinking logic, and of course makes it available to users who *don't* know they may have latent leap-second bugs
- I will *definitely* be calling you in to review leap-second information related code... we'll probably ship our own literal list as a last resort, but prefer to load from the OS or `tzdata` package (and maybe even `pytz`) if available.